src/lib/core/data-table/data-table.ts

```diff
@@ -99,34 +103,44 @@ export class CdkTable implements CollectionViewer {
 
   ngOnDestroy() {
     // TODO(andrewseguin): Disconnect from the data source so
-    // that it can unsubscribe from its streams.
+    //   that it can unsubscribe from its streams.
   }
 
   ngOnInit() {
     // TODO(andrewseguin): Setup a listener for scroll events
-    //   and emit the calculated view to this.viewChanged
+    //   and emit the calculated view to this.viewChange
   }
 
   ngAfterContentInit() {
     // TODO(andrewseguin): Throw an error if two columns share the same name
     this._columnDefinitions.forEach(columnDef => {
       this._columnDefinitionsByName.set(columnDef.name, columnDef);
     });
+
+    // Get and merge the streams for column changes made to the row defs
+    const rowDefs = [...this._rowDefinitions.toArray(), this._headerDefinition];
+    const columnChangeStreams = rowDefs.map((rowDef: BaseRowDef) => rowDef.columnsChange);
+    this.columnsChange = Observable.combineLatest(columnChangeStreams);
   }
 
   ngAfterViewInit() {
-    // TODO(andrewseguin): Re-render the header when the header's columns change.
     this.renderHeaderRow();
 
-    // TODO(andrewseguin): Re-render rows when their list of columns change.
+    // Re-render the header row if the columns changed.
+    this.columnsChange.subscribe(() => {
+      this._headerRowPlaceholder.viewContainer.clear();
+      this.renderHeaderRow();
+    });
+
     // TODO(andrewseguin): If the data source is not
     //   present after view init, connect it when it is defined.
     // TODO(andrewseguin): Unsubscribe from this on destroy.
-    this.dataSource.connect(this).subscribe((rowsData: any[]) => {
+    const renderChange = [this.dataSource.connect(this), this.columnsChange];
+    Observable.combineLatest(renderChange).subscribe(([data]) => {
```

Comment:
The only part of the observable we want is the first element in the array result, so just grabbing it here. If it looks too odd I can change it back to (result: any[]) and then grab result[0] for our data.